kj-rs/tests/test_maybe.rs

```diff
@@ -0,0 +1,148 @@
+use crate::ffi::{OpaqueCxxClass, Shared};
+use kj_rs::repr::{Maybe, Own};
+
+pub fn take_maybe_own_ret(val: Maybe<Own<OpaqueCxxClass>>) -> Maybe<Own<OpaqueCxxClass>> {
+    let mut option: Option<Own<OpaqueCxxClass>> = val.into();
+    if let Some(val) = &mut option {
+        val.as_mut().set_data(42);
+    }
+
+    option.into()
+}
+
+pub fn take_maybe_own(val: Maybe<Own<OpaqueCxxClass>>) {
+    let option: Option<Own<OpaqueCxxClass>> = val.into();
+    // Own gets destoyed at end of `if let` block, because it takes ownership of `option`
+    if let Some(own) = option {
+        assert_eq!(own.get_data(), 42);
+    }
+}
+
+/// # Safety: Uses a reference in a function that can be called from C++, which is opaque
+/// to the Rust compiler, so it cannot verify lifetime requirements
+pub unsafe fn take_maybe_ref_ret<'a>(val: Maybe<&'a u64>) -> Maybe<&'a u64> {
+    let option: Option<&u64> = val.into();
+    if let Some(num) = &option {
+        assert_eq!(**num, 15);
+    }
+    option.into()
+}
+
+pub fn take_maybe_ref(val: Maybe<&u64>) {
+    let mut option: Option<&u64> = val.into();
+    // Pure Rust at this point, but just in case
+    if let Some(val) = option.take() {
+        assert_eq!(*val, 15);
+    }
+}
+
+pub fn take_maybe_shared_ret(val: Maybe<Shared>) -> Maybe<Shared> {
+    let mut option: Option<Shared> = val.into();
+    if let Some(mut shared) = option.take() {
+        shared.i = 18;
+    }
+    option.into()
+}
+
+pub fn take_maybe_shared(val: Maybe<Shared>) {
+    let _: Option<Shared> = val.into();
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::ffi::{self, OpaqueCxxClass, Shared};
+    use kj_rs::repr::{Maybe, Own};
+
+    #[test]
+    fn test_some() {
+        let maybe: Maybe<i64> = ffi::return_maybe();
+        assert!(!maybe.is_none());
+    }
+
+    #[test]
+    fn test_none() {
+        let maybe: Maybe<i64> = ffi::return_maybe_none();
+        assert!(maybe.is_none());
+    }
+
+    #[test]
+    fn test_none_ptr() {
+        let maybe = ffi::return_maybe_ptr_none();
+        assert!(maybe.is_none());
+        assert!(!maybe.is_some());
+    }
+
+    #[test]
+    fn test_none_ptr_opt() {
+        let maybe = ffi::return_maybe_ptr_none();
+        let maybe: Option<*const i64> = maybe.into();
+        assert!(maybe.is_none());
+    }
+
+    #[test]
+    fn test_some_ptr() {
+        let maybe = ffi::return_maybe_ptr_some();
+        assert!(maybe.is_some());
+        assert!(!maybe.is_none());
```

Comment:
Can we also assert that the pointed-to-value is indeed 14? I have a suspicion it won't be, due to my observation that there's no actual `Maybe<T*>` specialization.